sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,330 @@
+    const dif_rv_timer_t *timer, uint32_t hart_id,
+    dif_rv_timer_enabled_t state) {
+  if (timer == NULL || hart_id >= timer->config.hart_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  switch (state) {
+    case kDifRvTimerEnabled:
+      mmio_region_nonatomic_set_bit32(timer->base_addr,
+                                      RV_TIMER_CTRL_REG_OFFSET, hart_id);
+      break;
+    case kDifRvTimerDisabled:
+      mmio_region_nonatomic_clear_bit32(timer->base_addr,
+                                        RV_TIMER_CTRL_REG_OFFSET, hart_id);
+      break;
+    default:
+      return kDifRvTimerBadArg;
+  }
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_counter_read(const dif_rv_timer_t *timer,
+                                                uint32_t hart_id,
+                                                uint64_t *out) {
+  if (timer == NULL || out == NULL || hart_id >= timer->config.hart_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  while (true) {
+    uint32_t upper = mmio_region_read32(
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_UPPER0_REG_OFFSET));
+    uint32_t lower = mmio_region_read32(
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_LOWER0_REG_OFFSET));
+
+    // Read the upper half again, to check that the lower half didn't
+    // overflow. If it did, we need to retry the read.
+    //
+    // Note that, if the counter is stopped, this check is guaranteed
+    // to always succeed.
+    uint32_t overflow_check = mmio_region_read32(
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_LOWER0_REG_OFFSET));
+
+    if (upper == overflow_check) {
+      *out = (((uint64_t)upper) << 32) | lower;
+      return kDifRvTimerOk;
+    }
+  }
+}
+
+dif_rv_timer_result_t dif_rv_timer_arm(const dif_rv_timer_t *timer,
+                                       uint32_t hart_id, uint32_t comp_id,
+                                       uint64_t threshold) {
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  uint32_t lower = threshold;
+  uint32_t upper = threshold >> 32;
+
+  ptrdiff_t lower_reg =
+      reg_for_hart(hart_id, RV_TIMER_COMPARE_LOWER0_0_REG_OFFSET) +
+      (sizeof(uint64_t) * comp_id);
+  ptrdiff_t upper_reg =
+      reg_for_hart(hart_id, RV_TIMER_COMPARE_UPPER0_0_REG_OFFSET) +
+      (sizeof(uint64_t) * comp_id);
+
+  // First, set the upper register to the largest value possible without setting
+  // off the alarm; this way, we can increase the lower register without setting
+  // off the alarm.
+  mmio_region_write32(timer->base_addr, upper_reg, UINT32_MAX);
+
+  // This can't set off the alarm because of the value we set above.
+  mmio_region_write32(timer->base_addr, lower_reg, lower);
+  // Finish writing the new value; this may set off an alarm immediately.
+  mmio_region_write32(timer->base_addr, upper_reg, upper);
```

Comment:
I don't think this makes a difference, but just wanted to point out that the risc-v spec has a sample that goes `0xFF` -> lower, set upper, set lower.

Also, in the first comment `increase` -> `set`.

Reply:
Yeah. I don't think it matters. You just need to set the timer to a known point in the future that you're unlikely to overflow while changing the two words.